src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,763 @@
+		if (gre_con == NULL) {
+			NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+					  "gre connection not found: client call_id %hu\n",
+					  pptp_call_reply->peer_call_id);
+			return false;
+		}
+
+		/* rewrite Peer Call ID */
+		old_call_id = pptp_call_reply->peer_call_id;
+		pptp_call_reply->peer_call_id = gre_con->orig_client_call_id;
+		/* save server call id */
+		gre_con->ctx.server_call_id = pptp_call_reply->hdr.call_id;
+
+		/* fix checksum */
+		cksum = tcp->check;
+		tcp->check = npf_fixup16_cksum(cksum, old_call_id,
+				  pptp_call_reply->peer_call_id);
+
+		/* if client and server call ids have been seen,
+		 * create new gre connection state entry
+		 */
+		if (gre_con->ctx.client_call_id != 0 &&
+				  gre_con->ctx.server_call_id != 0 &&
+				  gre_con->orig_client_call_id != 0) {
+			/* create pptp gre context cache */
+			memcpy(&gre_npc, npc, sizeof(npf_cache_t));
+			gre_npc.npc_proto = IPPROTO_GRE;
+			gre_npc.npc_info = NPC_IP46 | NPC_LAYER4 | NPC_ALG_PPTP_GRE_CTX;
+			gre_npc.npc_l4.pptp_gre_ctx = &gre_con->ctx;
+			/* setup ip addresses */
+			npf_nat_getorig(nt, &o_addr, &o_port);
+			gre_npc.npc_ips[NPF_SRC] = o_addr;
+			gre_npc.npc_ips[NPF_DST] = npc->npc_ips[NPF_SRC];
+			/* establish gre connection state and associate nat */
+			npfa_pptp_gre_establish_gre_conn(&gre_npc, PFIL_OUT, gre_con, nt);
+		}
+		break;
+
+	case PPTP_CALL_DISCONNECT_NOTIFY:
+		if (pptp->len < sizeof(struct pptp_msg_hdr))
+			return false;
+		npf_nat_getorig(nt, &o_addr, &o_port);
+
+		/* lookup for a gre connection entry */
+		gre_con = npfa_pptp_alg_arg_lookup_with_server_call_id(alg_arg,
+				  pptp->call_id);
+		if (gre_con == NULL)
+			return false;
+
+		npfa_pptp_expire_gre_con(npc->npc_ctx, gre_con, o_addr->word32[0],
+				  npc->npc_ips[NPF_SRC]->word32[0]);
+		break;
+
+	case PPTP_WAN_ERROR_NOTIFY:
+		if (pptp->len < sizeof(struct pptp_msg_hdr))
+			return false;
+
+		NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+				  "pptp wan error notify received, peer's call_id %hu\n",
+				  pptp->call_id);
+
+		gre_con = npfa_pptp_alg_arg_lookup_with_client_call_id(alg_arg,
+				  pptp->call_id);
+		if (gre_con == NULL)
+			return false;
+
+		old_call_id = pptp->call_id;
+		pptp->call_id = gre_con->orig_client_call_id;
+		cksum = tcp->check;
+		tcp->check = npf_fixup16_cksum(cksum, old_call_id,
+				  gre_con->orig_client_call_id);
+		break;
+
+	default:
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ *
+ */
+static void
+npfa_pptp_tcp_destroy(npf_t *npf, npf_conn_t *con)
+{
+	struct pptp_gre_con *gre_con;
+	struct pptp_alg_arg *alg_arg;
+	npf_connkey_t *fw;
+	uint32_t client_ip, server_ip;
+	int i;
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp tcp alg destroy a tcp connection %p\n", con);
+
+	alg_arg = (struct pptp_alg_arg *)npf_nat_get_alg_arg(con->c_nat);
+	fw = npf_conn_getforwkey(con);
+
+	/* only ipv4 is supported */
+	if (NPF_CONNKEY_ALEN(fw) == sizeof(uint32_t) || alg_arg == NULL)
+		return;
+
+	client_ip = fw->ck_key[2];
```

Comment:
There is a private `connkey_getkey()` function. Instead of poking the connection key, please use it (rename to `npf_connkey_getkey()` and make it public).